python/pyspark/sql/pandas/group_ops.py

```diff
@@ -501,7 +502,47 @@ def transformWithStateUDF(
                 )
 
             statefulProcessorApiClient.set_implicit_key(key)
-            result = statefulProcessor.handleInputRows(key, inputRows)
+
+            batch_timestamp = statefulProcessorApiClient.get_batch_timestamp()
+            watermark_timestamp = statefulProcessorApiClient.get_watermark_timestamp()
+            # process with invalid expiry timer info and emit data rows
+            data_iter = statefulProcessor.handleInputRows(
+                key, inputRows, TimerValues(batch_timestamp, watermark_timestamp), ExpiredTimerInfo(False))
+            statefulProcessorApiClient.set_handle_state(
+                StatefulProcessorHandleState.DATA_PROCESSED
+            )
+
+            if timeMode == "processingtime":
+                expiry_list: list[(any, int)] = statefulProcessorApiClient.get_expiry_timers(batch_timestamp)
+            elif timeMode == "eventtime":
+                expiry_list: list[(any, int)] = statefulProcessorApiClient.get_expiry_timers(watermark_timestamp)
+            else:
+                expiry_list = []
+
+            result_iter_list = []
+            # process with valid expiry time info and with empty input rows,
+            # only timer related rows will be emitted
+            for key_obj, expiry_timestamp in expiry_list:
+                if timeMode == "processingtime" and expiry_timestamp < batch_timestamp:
+                    result_iter_list.append(statefulProcessor.handleInputRows(
+                        (key_obj,), iter([]),
+                        TimerValues(batch_timestamp, watermark_timestamp),
```

Comment:
is `watermark_timestamp` needed for processingTime time mode and vise versa?

Reply:
Same above, this is to couple with behavior on Scala side.